tests/js/client/shell/shell-failures-communication-cluster.js

```diff
@@ -0,0 +1,97 @@
+/*jshint globalstrict:false, strict:false, maxlen: 400 */
+/*global fail, assertEqual, assertTrue, assertFalse, print */
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief test failure scenarios
+///
+/// @file
+///
+/// DISCLAIMER
+///
+/// Copyright 2021-2021 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Max Neunhoeffer
+/// @author Copyright 2021, ArangoDB GmbH, Cologne, Germany
+///////////////////////////////////////////////////////////////////////////////
+
+const jsunity = require("jsunity");
+const arangodb = require("@arangodb");
+const db = arangodb.db;
+const internal = require("internal");
+
+function shellCommunicationsFailureSuite () {
+  'use strict';
+  const cn = "UnitTestsAhuacatlCommFailures";
+
+  return {
+    setUpAll: function () {
+      internal.debugClearFailAt();
+      db._drop(cn);
+      db._create(cn, { numberOfShards: 3 });
+    },
+
+    setUp: function () {
+      internal.debugClearFailAt();
+    },
+
+    tearDownAll: function () {
+      internal.debugClearFailAt();
+      db._drop(cn);
+    },
+
+    tearDown: function () {
+      internal.debugClearFailAt();
+    },
+
+    testRetryOnLeaderRefusal: function() {
```

Comment:
This test is good.
However it is rather complicated to understand what it actually tests (see inline comment suggestions)

Also we now need a replica of this test for all our CRUD Apis if we want to test the behavior this way.

Reply:
I accept your suggestions. However, I do not think we need to replicate this test for all CRUD APIs, since we are only testing the behaviour of the NetworkFeature, and this is used in the same way for all CRUD APIs. This test does **not** test the behaviour of insert. It merely tests something which happens to be used by insert (and other APIs). Therefore I think it is good enough to run the test for insert.